src/Git/Pack/Index.hs

```diff
@@ -42,190 +39,151 @@ instance Show GitError where
-
 data PackIndexState
   = PackIndexStateV1
-    { pisHandle :: Handle
-    , pisTotalRecords :: Maybe Word32
+    { pisMmap :: MmapHandle
     , pisFanOutTable :: Map Word8 Word32
     , pisRecordNos :: Map Sha1 Word32
     , pisV1RecordOffsets :: Map Sha1 Word32
     , pisPackFileSha1 :: Maybe Sha1
     , pisSha1 :: Maybe Sha1}
   | PackIndexStateV2
-    { pisHandle :: Handle
-    , pisTotalRecords :: Maybe Word32
+    { pisMmap :: MmapHandle
     , pisFanOutTable :: Map Word8 Word32
     , pisRecordNos :: Map Sha1 Word32
     , pisRecordCrcs :: Map Sha1 Word32
     , pisV2RecordOffsets :: Map Sha1 Word64
     , pisPackFileSha1 :: Maybe Sha1
     , pisSha1 :: Maybe Sha1}
 
-packIndexState :: Version -> Handle -> PackIndexState
+packIndexState :: Version -> MmapHandle -> PackIndexState
 packIndexState v h = case v of
   Version1 -> PackIndexStateV1
-    h Nothing mempty mempty mempty Nothing Nothing
+    h mempty mempty mempty Nothing Nothing
   Version2 -> PackIndexStateV2
-    h Nothing mempty mempty mempty mempty Nothing Nothing
+    h mempty mempty mempty mempty Nothing Nothing
 
 pisVersion :: PackIndexState -> Version
 pisVersion PackIndexStateV1 {} = Version1
 pisVersion PackIndexStateV2 {} = Version2
 
 instance Show PackIndexState where
-  show pis = let v = pisVersion pis in case pisTotalRecords pis of
-    Nothing -> printf "<PackIndexState: %s>" (show v)
-    Just totRecs -> printf "<PackIndexState: %s, records: %d>" (show v) totRecs
+  show pis =
+    let
+      v = pisVersion pis
+      totRecs = runIdentity $ evalStateT getPackIndexTotalRecords pis
+    in
+      printf "<PackIndexState: %s, records: %d>" (show v) totRecs
 
 withPackIndex
   :: MonadIO m
-  => FilePath -> IOMode -> StateT PackIndexState (ExceptT GitError m) r
+  => FilePath -> StateT PackIndexState (ExceptT GitError m) r
   -> m (Either GitError r)
-withPackIndex path mode m = runExceptT $ do
-  h <- excTIO $ openBinaryFile path mode
+withPackIndex path m = runExceptT $ do
+  h <- excTIO $ mmapFileForeignPtr path ReadOnly Nothing
   v <- getPackIndexVersion h
   res <- evalStateT m (packIndexState v h)
-  -- FIXME: this might be better bracketed in case of error, although the RTS
-  -- should clean up the handle if it drops out of scope:
-  excTIO $ hClose h
   return res
 
-packIndexHeaderSize :: Version -> Integer
+packIndexHeaderSize :: Version -> Int
 packIndexHeaderSize Version1 = 0
 packIndexHeaderSize Version2 = 8
 
-packIndexDataStart :: Version -> Integer
+packIndexDataStart :: Version -> Int
 packIndexDataStart v = packIndexHeaderSize v + 256 * 4
 
-getPackIndexVersion :: (MonadIO m, MonadError GitError m) => Handle -> m Version
-getPackIndexVersion h = do
-  excTIO $ hSeek h AbsoluteSeek 0
-  (magic, rest) <- BS.splitAt 4 <$> excTIO (BS.hGet h 8)
-  version <- bsToWord32 rest
+getPackIndexVersion
+  :: MonadError GitError m => MmapHandle -> m Version
+getPackIndexVersion h = let magic = mmapData h (FromStart 0) (Length 4) in do
```

Comment:
Given it is used once in a fairly short line, not sure what the let here is buying.

Reply:
Yeah - it just used to be more complicated because it was doing IO to read it before. I'll inline.